src/clj/athens/self_hosted/components/web.clj

```diff
@@ -0,0 +1,118 @@
+    [compojure.core                    :as compojure]
+    [org.httpkit.server                :as http]))
+
+
+;; WebSocket handlers
+(defn open-handler
+  [channel]
+  (clients/add-client! channel))
+
+
+(defn close-handler
+  [channel status]
+  (let [username            (clients/get-client-username channel)
+        presence-disconnect {:presence {:username   username
+                                        :disconnect true}}]
+    (clients/remove-client! channel)
+    (log/info channel username "closed, status" status)
+    (when username
+      (clients/broadcast! presence-disconnect))))
+
+
+(defn- make-receive-handler
+  [datahike]
+  (fn receive-handler
+    [channel msg]
+    (log/debug channel "<-" msg)
+    (let [username (clients/get-client-username channel)
+          data     (clients/<-transit msg)]
+      (if-not (schema/valid-event? data)
+        (let [explanation (schema/explain-event data)]
+          (log/warn channel "invalid event received:" explanation)
+          (clients/send! channel {:event/id      (:event/id data)
+                                  :event/status  :rejected
+                                  :reject/reason explanation}))
+        (do
+          (log/debug channel "decoded event" (pr-str data))
+          (if (and (nil? username)
+                   (not= :presence/hello (:event/type data)))
+            (do
+              (log/warn channel "Message out of order, didn't say :presence/hello.")
+              (clients/send! channel {:event/id      (:event/id data)
+                                      :event/status  :rejected
+                                      :reject/reason :introduce-yourself}))
+            (let [event-type (:event/type data)
+                  result     (cond
+                               (contains? presence/supported-event-types event-type)
+                               (presence/presence-handler datahike channel data)
+
+                               (contains? datascript/supported-event-types event-type)
+                               (datascript/datascript-handler datahike channel data))]
+              (clients/send! channel (merge {:event/id (:event/id data)}
+                                            result)))))))))
+
+
+(defn- make-websocket-handler
+  [datahike]
+  (fn websocket-handler
+    [request]
+    (http/as-channel request
+                     {:on-open    #'open-handler
+                      :on-close   #'close-handler
+                      :on-receive (make-receive-handler datahike)})))
+
+
+(defn- make-ws-route
+  [datahike]
+  (compojure/routes
+    (compojure/GET "/ws" []
+                   (make-websocket-handler datahike))))
+
+
+(compojure/defroutes health-check-route
+                     (compojure/GET "/health-check" [] "ok"))
+
+
+(defn make-handler
+  [datahike]
+  (compojure/routes health-check-route
+                    (make-ws-route datahike)))
+
+
+(defrecord WebServer
+  [config httpkit datahike]
```

Comment:
Why is this ordered `[config httpkit datahike]`? I would expect this to be ordered `[config datahike httpkit]`? because config and datahike are dependencies and httpkit is added in constructor.

Reply:
`defrecord` semantics destructure the map based on key name.

Reply:
Order doesn't matter, but it does make sense to order meta-data first (I'm right :))